File: mexc_monitor/filters.py
from __future__ import annotations

import pandas as pd


def quote_suffix_for_filter(market: str, raw: str) -> str:
    s = raw.strip().upper()
    if not s:
        return "USDT" if market == "spot" else "_USDT"
    if market == "futures" and not s.startswith("_"):
        return f"_{s}"
    return s


def apply_cross_market_filters(
    df: pd.DataFrame,
    *,
    quote_raw: str,
    min_basis_bps: float,
    min_vol_quote: float,
    search: str,
    sort_by: str,
    ascending: bool,
) -> pd.DataFrame:
    """Фильтры для снимка спот↔фьючерс: |basis_mid_bps|, объёмы на обеих ногах."""
    quote_suffix = quote_suffix_for_filter("spot", quote_raw)
    view = df.copy()
    if quote_suffix:
        view = view[view["symbol_spot"].str.endswith(quote_suffix, na=False)]
    if min_basis_bps > 0:
        view = view[view["basis_mid_bps"].fillna(0).abs() >= min_basis_bps]
    if min_vol_quote > 0:
        view = view[
            (view["volume_24h_quote_spot"].fillna(0) >= min_vol_quote)
            & (view["volume_24h_quote_fut"].fillna(0) >= min_vol_quote)
        ]
    if search:
        su = search.strip().upper()
        view = view[
            view["symbol_spot"].str.contains(su, na=False)
            | view["symbol_futures"].str.contains(su, na=False)
        ]
    sort_col = (
        sort_by
        if sort_by in view.columns
        else (
            "basis_mid_bps"
            if "basis_mid_bps" in view.columns
            else ("symbol_spot" if "symbol_spot" in view.columns else sort_by)
        )
    )
    if sort_col in view.columns:
        na_pos = "last" if ascending else "first"
        view = view.sort_values(by=sort_col, ascending=ascending, na_position=na_pos)
    return view


def apply_market_filters(
    df: pd.DataFrame,
    *,
    market: str,
    quote_raw: str,
    min_spread_bps: float,
    min_vol_quote: float,
    search: str,
    sort_by: str,
    ascending: bool,
    min_bid_l1_notional_quote: float = 0.0,
    min_ask_l1_notional_quote: float = 0.0,
) -> pd.DataFrame:
    """Те же правила, что в Streamlit UI (фильтрация и сортировка локально по снимку)."""
    if market == "cross":
        return apply_cross_market_filters(
            df,
            quote_raw=quote_raw,
            min_basis_bps=min_spread_bps,
            min_vol_quote=min_vol_quote,
            search=search,
            sort_by=sort_by,
            ascending=ascending,
        )
    quote_suffix = quote_suffix_for_filter(market, quote_raw)
    view = df.copy()
    if quote_suffix:
        view = view[view["symbol"].str.endswith(quote_suffix, na=False)]
    if min_spread_bps > 0:
        view = view[view["spread_bps"].fillna(0) >= min_spread_bps]
    if min_vol_quote > 0:
        view = view[view["volume_24h_quote"].fillna(0) >= min_vol_quote]
    if min_bid_l1_notional_quote > 0 and "bid_qty" in view.columns and "bid" in view.columns:
        bn = view["bid_qty"].fillna(0) * view["bid"].fillna(0)
        view = view[bn >= min_bid_l1_notional_quote]
    if min_ask_l1_notional_quote > 0 and "ask_qty" in view.columns and "ask" in view.columns:
        an = view["ask_qty"].fillna(0) * view["ask"].fillna(0)
        view = view[an >= min_ask_l1_notional_quote]
    if search:
        view = view[view["symbol"].str.contains(search.strip().upper(), na=False)]

    sort_col = (
        sort_by
        if sort_by in view.columns
        else (
            "spread_bps"
            if "spread_bps" in view.columns
            else ("symbol" if "symbol" in view.columns else sort_by)
        )
    )
    if sort_col in view.columns:
        na_pos = "last" if ascending else "first"
        view = view.sort_values(by=sort_col, ascending=ascending, na_position=na_pos)
    return view

File: frontend/src/filters.ts
/**
 * Дублирует логику mexc_monitor/filters.py для клиентской фильтрации без повторных запросов к API.
 */
import type {
  CrossMarketRow,
  Exchange,
  Market,
  MarketRow,
  SnapshotRow,
} from "./types";
import { EXCHANGE_DEFAULT_QUOTE } from "./types";

const UI_LOG = import.meta.env.DEV;

export function isCrossMarketRow(r: unknown): r is CrossMarketRow {
  if (!r || typeof r !== "object") return false;
  const o = r as Record<string, unknown>;
  return (
    typeof o.symbol_spot === "string" &&
    typeof o.symbol_futures === "string"
  );
}

export function quoteSuffixForFilter(
  market: Market,
  raw: string,
  exchange: Exchange = "mexc",
): string {
  const s = String(raw ?? "")
    .trim()
    .toUpperCase();
  // Подчёркнутый формат символа (BTC_USDT) — только у MEXC futures.
  const mexcFutures = exchange === "mexc" && market === "futures";
  if (!s) {
    if (mexcFutures) return "_USDT";
    return EXCHANGE_DEFAULT_QUOTE[exchange] ?? "USDT";
  }
  if (mexcFutures && !s.startsWith("_")) return `_${s}`;
  return s;
}

export interface FilterOptions {
  market: Market;
  /** Биржа снимка; влияет на дефолтный суффикс котировки. По умолчанию MEXC. */
  exchange?: Exchange;
  quoteRaw: string;
  minSpreadBps: number;
  minVolQuote: number;
  search: string;
  sortBy: string;
  ascending: boolean;
  /**
   * Мин. нотация на лучшем bid (bid_qty × bid), USDT≈. 0 — не фильтровать.
   * Для режима отбора пар под сбор спреда с «плотностью» на touch bid.
   */
  minBidL1NotionalQuote: number;
  /** Мин. нотация на лучшем ask (ask_qty × ask), USDT≈. */
  minAskL1NotionalQuote: number;
}

function num(v: number | null | undefined): number {
  if (v == null || Number.isNaN(v)) return 0;
  return v;
}

function effectiveCrossSortKey(
  rows: CrossMarketRow[],
  sortBy: string,
): keyof CrossMarketRow {
  if (!rows.length) return "basis_mid_bps";
  const k = sortBy as keyof CrossMarketRow;
  if (k in rows[0]) return k;
  return "basis_mid_bps";
}

function effectiveSpotFutSortKey(
  rows: MarketRow[],
  sortBy: string,
): keyof MarketRow {
  if (!rows.length) return "spread_bps";
  const k = sortBy as keyof MarketRow;
  if (k in rows[0]) return k;
  return "spread_bps";
}

export function applyCrossMarketFilters(
  rows: CrossMarketRow[],
  o: FilterOptions,
): CrossMarketRow[] {
  const suffix = quoteSuffixForFilter("spot", o.quoteRaw);
  const search = String(o.search ?? "")
    .trim()
    .toUpperCase();

  let out = rows.filter((r) => {
    const spot = (r.symbol_spot ?? "").toUpperCase();
    const fut = (r.symbol_futures ?? "").toUpperCase();
    if (suffix && !spot.endsWith(suffix)) return false;
    if (o.minSpreadBps > 0 && Math.abs(num(r.basis_mid_bps)) < o.minSpreadBps)
      return false;
    if (
      o.minVolQuote > 0 &&
      (num(r.volume_24h_quote_spot) < o.minVolQuote ||
        num(r.volume_24h_quote_fut) < o.minVolQuote)
    )
      return false;
    if (search && !spot.includes(search) && !fut.includes(search)) return false;
    return true;
  });

  const key = effectiveCrossSortKey(out, o.sortBy);
  const dir = o.ascending ? 1 : -1;

  out = [...out].sort((a, b) => {
    const av = a[key];
    const bv = b[key];
    if (av == null && bv == null) return 0;
    if (av == null) return o.ascending ? 1 : -1;
    if (bv == null) return o.ascending ? -1 : 1;
    if (typeof av === "number" && typeof bv === "number") {
      if (av === bv)
        return String(a.symbol_spot ?? "").localeCompare(
          String(b.symbol_spot ?? ""),
        );
      return av < bv ? -dir : dir;
    }
    const as = String(av);
    const bs = String(bv);
    if (as === bs) return 0;
    return as < bs ? -dir : dir;
  });

  return out;
}

export function applyMarketFilters(
  rows: SnapshotRow[],
  o: FilterOptions,
): SnapshotRow[] {
  if (o.market === "cross") {
    const valid = rows.filter(isCrossMarketRow);
    if (valid.length !== rows.length && rows.length > 0 && UI_LOG) {
      console.warn(
        `[MEXC UI] cross: отброшено ${rows.length - valid.length} строк (нет symbol_spot/symbol_futures)`,
      );
    }
    return applyCrossMarketFilters(valid, o);
  }

  const suffix = quoteSuffixForFilter(o.market, o.quoteRaw, o.exchange);
  const search = String(o.search ?? "")
    .trim()
    .toUpperCase();

  let out = (rows as MarketRow[]).filter((r) => {
    const sym = String(r.symbol ?? "")
      .trim()
      .toUpperCase();
    if (!sym) return false;
    if (suffix && !sym.endsWith(suffix)) return false;
    if (o.minSpreadBps > 0 && num(r.spread_bps) < o.minSpreadBps) return false;
    if (o.minVolQuote > 0 && num(r.volume_24h_quote) < o.minVolQuote)
      return false;
    if (o.minBidL1NotionalQuote > 0) {
      const bn = num(r.bid_qty) * num(r.bid);
      if (bn < o.minBidL1NotionalQuote) return false;
    }
    if (o.minAskL1NotionalQuote > 0) {
      const an = num(r.ask_qty) * num(r.ask);
      if (an < o.minAskL1NotionalQuote) return false;
    }
    if (search && !sym.includes(search)) return false;
    return true;
  });

  const key = effectiveSpotFutSortKey(out, o.sortBy);
  const dir = o.ascending ? 1 : -1;

  out = [...out].sort((a, b) => {
    const av = a[key];
    const bv = b[key];
    if (av == null && bv == null) return 0;
    if (av == null) return o.ascending ? 1 : -1;
    if (bv == null) return o.ascending ? -1 : 1;
    if (typeof av === "number" && typeof bv === "number") {
      if (av === bv)
        return String(a.symbol ?? "").localeCompare(String(b.symbol ?? ""));
      return av < bv ? -dir : dir;
    }
    const as = String(av);
    const bs = String(bv);
    return as === bs ? 0 : as < bs ? -dir : dir;
  });

  return out;
}
